### DISCUSSION TITLE
Loading speed on AtmelICE

### DISCUSSION THREAD

**Discussion by celephicus** [bloombloombloom/Bloom#94] [Q&A]
Bloom works well, and has let me retire my Windows laptop as the only reason I had it was for Atmel Studio debugging. But I'm using the Mega328P on DebugWire, and the upload speed is very slow. Is there any way of speeding it up a bit?

Tom

**Comment by navnavnav** [comment]
Could you please enable debug logging and send me the log output of a debug session where you run the `load` command? This will give me an idea of what's going on. You may have to put the log output in a text file and attach to the comment as GitHub has some character length restrictions for comments.

To enable debug logging, just set `debugLogging` to `true` in the root node of your bloom.yaml:

```yaml
debugLogging: true

environments:
...
```

Does Atmel Studio upload faster? To what extent?

Does your ELF contain EEPROM sections? IIRC, writing EEPROM is quite slow on debugWire targets, so that could be the problem.

What's the size of the program you're uploading?

Are you using any IDE like CLion? Is it restarting Bloom for every debug session? If so, it could be startup/shutdown times that are slowing things down. I [raised this with Jetbrains](https://youtrack.jetbrains.com/issue/CPP-35677/Add-option-to-maintain-active-GDB-connection-when-rebuilding-and-rerunning-embedded-program-for-better-faster-debugging) a few weeks back.

**Reply by navnavnav** [reply]
BTW, the debug log output may contain binary segments of your EEPROM memory, and maybe other sections of your ELF.

If this is a problem (you don't want to expose certain segments of your program), I can walk you through truncating the appropriate log lines, if you wish.

**Comment by celephicus** [comment]
Sorry to take so long, I've done some testing, removed the eeprom section which made little difference. My program is about20KB, quite modest. The bloom log is attached.
[bloom.log](https://github.com/bloombloombloom/Bloom/files/13537711/bloom.log)
 Sorry I haven't tried Atmel Studio yet, I can install it on another laptop if this is important, but it is significant effort.

**Comment by celephicus** [comment]
The target was an 8MHz 328P.I can confirm that loading similar size programs to the same target running at 16MHz takes a few seconds, not over 30 secs.

**Comment by navnavnav** [comment]
In v2.0.0, I have introduced support for delta programming, where Bloom, during a programming session, will compare the new program with the program memory cache to identify regions of memory that have changed, and then reduce the upload to only the changed regions.

If the binary hasn't changed much, this could significantly improve programming speeds.

Delta programming is supported on AVR UPDI, debugWIRE and RISC-V targets.

I'm closing this discussion now, as there is no more that I'm willing to do on this issue.
